File: frontend/lib/household.ts
/**
 * Build a PolicyEngine household situation for the PE API.
 *
 * For the WV SB 392 tax-cut dashboard:
 * - PolicyEngine baseline already reflects the SB 392 cut (post-cut rates).
 * - The "revert" policy reverts the 2026 rates back to 2025 values.
 * - The household calculator uses revert as the *baseline* and current law
 *   as the *reform* so impact = current_law - pre_cut (positive = household
 *   gains under the cut).
 */

import type { HouseholdRequest } from "./types";

const GROUP_UNITS = ["families", "spm_units", "tax_units", "households"] as const;

/**
 * Revert-reform policy: rolls 2026 WV income tax rates back to their 2025
 * values. Mirrors reform_revert.json so the household calculator ships
 * the same overrides used by Modal.
 */
const REVERT_POLICY: Record<string, Record<string, number>> = {
  "gov.states.wv.tax.income.rates.single[0].rate": { "2026-01-01.2100-12-31": 0.0222 },
  "gov.states.wv.tax.income.rates.single[1].rate": { "2026-01-01.2100-12-31": 0.0296 },
  "gov.states.wv.tax.income.rates.single[2].rate": { "2026-01-01.2100-12-31": 0.0333 },
  "gov.states.wv.tax.income.rates.single[3].rate": { "2026-01-01.2100-12-31": 0.0444 },
  "gov.states.wv.tax.income.rates.single[4].rate": { "2026-01-01.2100-12-31": 0.0482 },
  "gov.states.wv.tax.income.rates.joint[0].rate": { "2026-01-01.2100-12-31": 0.0222 },
  "gov.states.wv.tax.income.rates.joint[1].rate": { "2026-01-01.2100-12-31": 0.0296 },
  "gov.states.wv.tax.income.rates.joint[2].rate": { "2026-01-01.2100-12-31": 0.0333 },
  "gov.states.wv.tax.income.rates.joint[3].rate": { "2026-01-01.2100-12-31": 0.0444 },
  "gov.states.wv.tax.income.rates.joint[4].rate": { "2026-01-01.2100-12-31": 0.0482 },
  "gov.states.wv.tax.income.rates.head[0].rate": { "2026-01-01.2100-12-31": 0.0222 },
  "gov.states.wv.tax.income.rates.head[1].rate": { "2026-01-01.2100-12-31": 0.0296 },
  "gov.states.wv.tax.income.rates.head[2].rate": { "2026-01-01.2100-12-31": 0.0333 },
  "gov.states.wv.tax.income.rates.head[3].rate": { "2026-01-01.2100-12-31": 0.0444 },
  "gov.states.wv.tax.income.rates.head[4].rate": { "2026-01-01.2100-12-31": 0.0482 },
  "gov.states.wv.tax.income.rates.surviving_spouse[0].rate": { "2026-01-01.2100-12-31": 0.0222 },
  "gov.states.wv.tax.income.rates.surviving_spouse[1].rate": { "2026-01-01.2100-12-31": 0.0296 },
  "gov.states.wv.tax.income.rates.surviving_spouse[2].rate": { "2026-01-01.2100-12-31": 0.0333 },
  "gov.states.wv.tax.income.rates.surviving_spouse[3].rate": { "2026-01-01.2100-12-31": 0.0444 },
  "gov.states.wv.tax.income.rates.surviving_spouse[4].rate": { "2026-01-01.2100-12-31": 0.0482 },
  "gov.states.wv.tax.income.rates.separate[0].rate": { "2026-01-01.2100-12-31": 0.0222 },
  "gov.states.wv.tax.income.rates.separate[1].rate": { "2026-01-01.2100-12-31": 0.0296 },
  "gov.states.wv.tax.income.rates.separate[2].rate": { "2026-01-01.2100-12-31": 0.0333 },
  "gov.states.wv.tax.income.rates.separate[3].rate": { "2026-01-01.2100-12-31": 0.0444 },
  "gov.states.wv.tax.income.rates.separate[4].rate": { "2026-01-01.2100-12-31": 0.0482 },
};

function addMemberToUnits(
  situation: Record<string, unknown>,
  memberId: string
): void {
  for (const unit of GROUP_UNITS) {
    const unitObj = situation[unit] as Record<string, { members: string[] }>;
    const key = Object.keys(unitObj)[0];
    unitObj[key].members.push(memberId);
  }
}

export function buildHouseholdSituation(
  params: HouseholdRequest
): Record<string, unknown> {
  const {
    age_head,
    age_spouse,
    dependent_ages,
    income,
    year,
    max_earnings,
    state_code,
  } = params;
  const effectiveStateCode = state_code || "WV";
  const yearStr = String(year);
  const axisMax = Math.max(max_earnings, income);

  const situation: Record<string, unknown> = {
    people: {
      you: {
        age: { [yearStr]: age_head },
        employment_income: { [yearStr]: null },
      },
    },
    families: { "your family": { members: ["you"] } },
    marital_units: { "your marital unit": { members: ["you"] } },
    spm_units: { "your household": { members: ["you"] } },
    tax_units: {
      "your tax unit": {
        members: ["you"],
        adjusted_gross_income: { [yearStr]: null },
        income_tax: { [yearStr]: null },
        wv_income_tax: { [yearStr]: null },
      },
    },
    households: {
      "your household": {
        members: ["you"],
        state_code: { [yearStr]: effectiveStateCode },
        household_net_income: { [yearStr]: null },
      },
    },
    axes: [
      [
        {
          name: "employment_income",
          min: 0,
          max: axisMax,
          count: Math.min(4001, Math.max(501, Math.floor(axisMax / 500))),
          period: yearStr,
          target: "person",
        },
      ],
    ],
  };

  if (age_spouse != null) {
    const people = situation.people as Record<string, Record<string, unknown>>;
    people["your partner"] = { age: { [yearStr]: age_spouse } };
    addMemberToUnits(situation, "your partner");
    const maritalUnits = situation.marital_units as Record<string, { members: string[] }>;
    maritalUnits["your marital unit"].members.push("your partner");
  }

  for (let i = 0; i < dependent_ages.length; i++) {
    const childId =
      i === 0
        ? "your first dependent"
        : i === 1
          ? "your second dependent"
          : `dependent_${i + 1}`;

    const people = situation.people as Record<string, Record<string, unknown>>;
    people[childId] = { age: { [yearStr]: dependent_ages[i] } };
    addMemberToUnits(situation, childId);
    const maritalUnits = situation.marital_units as Record<string, { members: string[] }>;
    maritalUnits[`${childId}'s marital unit`] = {
      members: [childId],
    };
  }

  return situation;
}

/**
 * Build the revert policy dict for the PE API. Used by the household
 * calculator as the *baseline*: applying it rolls WV's 2026 rates back to
 * 2025 values, so impact = current_law - revert isolates the SB 392 cut.
 */
export function buildRevertPolicy(): Record<string, Record<string, number>> {
  return REVERT_POLICY;
}

/**
 * Linear interpolation helper - find the value at `x` in sorted arrays.
 */
export function interpolate(
  xs: number[],
  ys: number[],
  x: number
): number {
  if (x <= xs[0]) return ys[0];
  if (x >= xs[xs.length - 1]) return ys[ys.length - 1];
  for (let i = 1; i < xs.length; i++) {
    if (xs[i] >= x) {
      const t = (x - xs[i - 1]) / (xs[i] - xs[i - 1]);
      return ys[i - 1] + t * (ys[i] - ys[i - 1]);
    }
  }
  return ys[ys.length - 1];
}

File: nj_credit_calc/household.py
"""
Household situation builder for the 2026 Utah tax changes calculator.

This module builds PolicyEngine household situations used by the
household calculator view. The calculator contrasts current-law 2026
(baseline) with the pre-2026 (reverted) parameters described in
``reform.json``; ``impact = current_law - reverted``.
"""

from typing import Any, Dict, List, Optional
import numpy as np


GROUP_UNITS = ["families", "spm_units", "tax_units", "households"]


def _add_member_to_units(
    situation: Dict[str, Any], member_id: str
) -> None:
    """Add a member to all group units in the situation."""
    for unit in GROUP_UNITS:
        unit_dict = situation[unit]
        first_key = next(iter(unit_dict))
        unit_dict[first_key]["members"].append(member_id)


def build_household_situation(
    age_head: int,
    age_spouse: Optional[int],
    dependent_ages: List[int],
    income: float,
    year: int,
    max_earnings: float,
    state_code: str = "UT",
    include_axes: bool = True,
) -> Dict[str, Any]:
    """
    Build a PolicyEngine household situation dictionary.

    Args:
        age_head: Age of the primary filer
        age_spouse: Age of spouse (None if single)
        dependent_ages: List of ages for each dependent
        income: Employment income
        year: Tax year
        max_earnings: Maximum earnings for axis calculation
        state_code: Two-letter state code (default: "UT")
        include_axes: Whether to include axes for parametric calculation

    Returns:
        A household situation dictionary for the PolicyEngine API.
    """
    year_str = str(year)
    axis_max = max(max_earnings, income)

    situation: Dict[str, Any] = {
        "people": {
            "you": {
                "age": {year_str: age_head},
                "employment_income": {year_str: income if not include_axes else None},
            },
        },
        "families": {"your family": {"members": ["you"]}},
        "marital_units": {"your marital unit": {"members": ["you"]}},
        "spm_units": {"your household": {"members": ["you"]}},
        "tax_units": {
            "your tax unit": {
                "members": ["you"],
                "adjusted_gross_income": {year_str: None},
            },
        },
        "households": {
            "your household": {
                "members": ["you"],
                "state_code": {year_str: state_code},
                "household_net_income": {year_str: None},
            },
        },
    }

    if include_axes:
        count = min(4001, max(501, int(axis_max / 500)))
        situation["axes"] = [
            [
                {
                    "name": "employment_income",
                    "min": 0,
                    "max": axis_max,
                    "count": count,
                    "period": year_str,
                    "target": "person",
                },
            ],
        ]

    # Add spouse if married
    if age_spouse is not None:
        situation["people"]["your partner"] = {"age": {year_str: age_spouse}}
        _add_member_to_units(situation, "your partner")
        situation["marital_units"]["your marital unit"]["members"].append("your partner")

    # Add dependents
    for i, age in enumerate(dependent_ages):
        if i == 0:
            child_id = "your first dependent"
        elif i == 1:
            child_id = "your second dependent"
        else:
            child_id = f"dependent_{i + 1}"

        situation["people"][child_id] = {"age": {year_str: age}}
        _add_member_to_units(situation, child_id)
        situation["marital_units"][f"{child_id}'s marital unit"] = {
            "members": [child_id],
        }

    return situation


def calculate_household_impact(
    age_head: int,
    age_spouse: Optional[int],
    dependent_ages: List[int],
    income: float,
    year: int,
    max_earnings: float,
    state_code: str = "UT",
) -> Dict[str, Any]:
    """
    Calculate household impact of the 2026 Utah tax changes using PolicyEngine.

    This is the Python equivalent of the calculator view in the frontend.
    It runs two simulations — the current-law (PolicyEngine baseline) and
    the reverted-to-2025 (reform applied) case — and returns
    ``impact = baseline - reform``.

    Args:
        age_head: Age of the primary filer
        age_spouse: Age of spouse (None if single)
        dependent_ages: List of ages for each dependent
        income: Employment income
        year: Tax year
        max_earnings: Maximum earnings for chart x-axis
        state_code: Two-letter state code (default: "UT")

    Returns:
        Impact analysis results including income_range, net_income_change,
        ut_income_tax_change, and income_tax_change (federal).
    """
    # Import here so the module can be imported without policyengine_us
    try:
        from policyengine_us import Simulation
        from policyengine_core.reforms import Reform
    except ImportError:
        raise ImportError(
            "policyengine_us is required for calculate_household_impact. "
            "Install it with: pip install policyengine-us"
        )

    from .reforms import load_reform

    situation = build_household_situation(
        age_head=age_head,
        age_spouse=age_spouse,
        dependent_ages=dependent_ages,
        income=income,
        year=year,
        max_earnings=max_earnings,
        state_code=state_code,
        include_axes=True,
    )

    # Baseline = current 2026 law already merged into PolicyEngine-US
    baseline_sim = Simulation(situation=situation)
    baseline_net_income = baseline_sim.calculate("household_net_income", year)
    baseline_ut_income_tax = baseline_sim.calculate("ut_income_tax", year)
    baseline_income_tax = baseline_sim.calculate("income_tax", year)
    income_range = baseline_sim.calculate("employment_income", year)

    # Reform = reverted parameters (pre-2026 values from reform.json)
    reform = Reform.from_dict(load_reform(), country_id="us")
    reform_sim = Simulation(situation=situation, reform=reform)
    reform_net_income = reform_sim.calculate("household_net_income", year)
    reform_ut_income_tax = reform_sim.calculate("ut_income_tax", year)
    reform_income_tax = reform_sim.calculate("income_tax", year)

    # Impact is current-law minus reverted (flipped from the typical
    # reform - baseline formulation because our reform reverts policy).
    net_income_change = baseline_net_income - reform_net_income
    ut_income_tax_change = baseline_ut_income_tax - reform_ut_income_tax
    income_tax_change = baseline_income_tax - reform_income_tax

    def interpolate(xs: np.ndarray, ys: np.ndarray, x: float) -> float:
        if x <= xs[0]:
            return float(ys[0])
        if x >= xs[-1]:
            return float(ys[-1])
        return float(np.interp(x, xs, ys))

    baseline_at_income = interpolate(income_range, baseline_net_income, income)
    reform_at_income = interpolate(income_range, reform_net_income, income)

    return {
        "income_range": income_range.tolist(),
        "net_income_change": net_income_change.tolist(),
        "ut_income_tax_change": ut_income_tax_change.tolist(),
        "income_tax_change": income_tax_change.tolist(),
        "benefit_at_income": {
            "current_law": baseline_at_income,
            "reverted": reform_at_income,
            "difference": baseline_at_income - reform_at_income,
        },
        "x_axis_max": max_earnings,
    }
